### ISSUE TITLE
Obfuscates public constants; causing problems/exceptions

### ISSUE THREAD

**Issue by agendec-com** [mkaring/ConfuserEx#27]
It seems the contents of public constants are being obfuscated, causing problems/exceptions during app runtime.
In my case it seems the "dpConfig" in a class level constant I have is being obfuscated:
Public Const DEFAULT_ROOT_ELEMENT_NAME As String = "dpConfig"

As I've understood from the issue below, such constants should not be obfuscated?
https://github.com/yck1509/ConfuserEx/issues/369

**Comment by mkaring** [comment]
The constants itself aren't encrypted. How ever the usage of the constants doesn't contain a reference to the original constant definition anymore. There is just the instruction to load the string "dpConfig" left in the end. And this instruction is encrypted. So you will find the
```VB.NET
Public Const DEFAULT_ROOT_ELEMENT_NAME As String = "dpConfig"
```
left in the definition of the class unchanged. But the usages of this constant will be encrypted.

**Comment by agendec-com** [comment]
Ok, but then I assume something else is wrong with the obfuscation since when I use that const as part of a string input to XmlDocument.LoadXml I get the following exception:
"System.Xml.XmlException: Tegnet >, heksadesimalverdi 0x3E, kan ikke innlede et navn. Linje 2, posisjon 2."
English translation:
"System.Xml.XmlException: The character >, hexadecimal value 0x3E, cannot initiate a name. Line 2, position 2."

Attached is the code (DEFAULT_ROOT_ELEMENT_NAME is assigned to the psRootElement input param).
[source.txt](https://github.com/mkaring/ConfuserEx/files/2943759/source.txt)

Non obfuscated version works ok.

**Comment by mkaring** [comment]
Using the exact code you provided, it works without problems for me. The error message indicates that the value passed to the constructor is not correct. How is the constructor invoked. Maybe something there is breaking the resource encryption?

**Comment by agendec-com** [comment]
Attached is the relevant part of the source - it's the constructor with no parameters calling the constructor with parameter.
[source.txt](https://github.com/mkaring/ConfuserEx/files/2944751/source.txt)

**Comment by mkaring** [comment]
I am using this exact code and it works for me. What version of Visual Studio or rather what version of Visual Basic are you using? Any specific configuration to the project? (One of the `Options` set to an non-default version)

**Comment by agendec-com** [comment]
I'm using Visual Basic 2015 - Visual Studio Enterprise 2015 ver. 14.0.25431.01 Update 3.
The solution consists of 2 projects - the startup project (DPAgenda) is an exe WinForms app, the other (AgFoundation) is a class library.
When I was going to look at the project settings I noticed something weird; the startup project did not have anything set in the "Startup object" setting. I don't know how this have happened; It could be Visual Studio 2017 which I tried using on the solution some time ago but it was quite buggy so I switched back to VS 2015.
Anyhow, I then tried setting the Startup object to what it should be (Sub Main) and obfuscated, and then it worked!
So then I tried obfuscating again with the previous version where Startup object was not set, and the problem was back again.
So it seems that has caused this; could this be right?
Also, could this be the cause of issue 26 I've reported?

**Comment by mkaring** [comment]
That sounds like the problem. The obfuscator looks for the method that is marked as start method. It is a bit strange that it works without obfuscation. Means it is able to locate the start method some other way. I'll look into that. It's a good find!

**Comment by mkaring** [comment]
I just tried to replicate the issue. How ever in case I am trying to set the startup object of an windows forms application to none, it always gives me a [`Microsoft.VisualBasic.ApplicationServices.NoStartupFormException`][1]. I tried doing this by removing the Startup object from the `*.vbproj` file and from the `Application.myapp`. Either way I get the exception. When I try to use a normal main method, and remove the entry from the `*.vbproj` file it shows that there is no start object, but it still uses the existing `Main` method as entry point to the assembly. So the only real way to figure out what is wrong there is to know how it was setup before you changed the project file.

Can you determine what exactly was changed on the project when you fixed the startup object setting? Talking about the project files here.

[1]: https://docs.microsoft.com/dotnet/api/microsoft.visualbasic.applicationservices.nostartupformexception

**Comment by agendec-com** [comment]
Only the startup object was changed - see comparison in Notepad++ below.
I've also attached the proj file that was missing the startup object.
It's probably worth noting that this project is very old; the first version of it I think was created in year 2003 so it's been converted through multiple VS and .Net framework versions.

![image](https://user-images.githubusercontent.com/45871299/54416032-792adb00-46fe-11e9-9f13-fe34c86fe481.png)
[DPAgenda.vbproj.txt](https://github.com/mkaring/ConfuserEx/files/2969935/DPAgenda.vbproj.txt)

**Comment by mkaring** [comment]
Even with your project file copied into my test project I am not able to reproduce the issue. :(
It will come down to the point that I need the unobfuscated assembly for analysis to locate the issue.

**Comment by agendec-com** [comment]
Unfortunately I can't give you the unobfuscated version of the assemblies.
But the situation that this occurs in is very obscure so I would assume you don't need to do anything about it?

**Comment by mkaring** [comment]
Another idea: Could you extract the unobfuscated IL instructions of the `dpConfig` class? So from the compiled assembly and open it with ILSpy or dnSpy and get the IL code.

Thats the last idea I got to identify the issue. If that doesn't work I'll drop the issue and hope that it doesn't occur in normal operation. :wink:

**Comment by agendec-com** [comment]
Ok, here it is, good luck 😊
[dpconfig from dnspy.txt](https://github.com/mkaring/ConfuserEx/files/2980501/dpconfig.from.dnspy.txt)

**Comment by mkaring** [comment]
That is actually the visual basic code. I'll need the IL instructions. You can switch to the IL instructions in the toolbar of dnspy in the drop down control that currently reads Basic.

**Comment by agendec-com** [comment]
Ops, sorry about that, here it is.
[dpconfig from dnspy.txt](https://github.com/mkaring/ConfuserEx/files/2980599/dpconfig.from.dnspy.txt)

**Comment by mkaring** [comment]
I can't reproduce this issue. I got an assembly with the exact same IL instructions as you got, still it works fine for me. Could you do a test with only the `constants` protection enabled, to identify if that protection is causing the issue?

**Comment by agendec-com** [comment]
If I only have constants enabled, the app can't startup - I get the System.IndexOutOfRangeException exception during startup.
I've attached the crproj file - is this correct?
[AgCliConfuserEx1.1.0.Test.crproj.txt](https://github.com/mkaring/ConfuserEx/files/3018691/AgCliConfuserEx1.1.0.Test.crproj.txt)

**Comment by mkaring** [comment]
Yes. Your crproj file is correct. Could you send me the IL instructions of the class after the obfuscation?  So I can see what the obfuscation is doing.

**Comment by agendec-com** [comment]
Here it is.
[dpconfig.from.dnspy.txt](https://github.com/mkaring/ConfuserEx/files/3020086/dpconfig.from.dnspy.txt)

**Comment by mkaring** [comment]
Okay. The reference to decoder is present. That points to the possibility that the decoder is not working properly. I'll look into that.

**Comment by agendec-com** [comment]
Ok, fingers crossed 🤞
